File: 207/207_version1.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        self.nodes = [[] for _ in range(numCourses)]
        self.visited = [0 for _ in range(numCourses)]
        for curr,pre in prerequisites:
            self.nodes[pre].append(curr)
        for n in range(numCourses):
            if self.dfs(n): return False
        return True
    def dfs(self,node):
        if self.visited[node]==1:
            return True
        if self.visited[node]==2:
            return False
        self.visited[node] = 1
        for i in self.nodes[node]:
            if self.dfs(i):return True
        self.visited[node] = 2
        return False

File: 207/207_version1.java
class Solution {
    private int[] visited;
    private List<ArrayList<Integer>> nodes;
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        if(numCourses==0) return true;
        nodes = new ArrayList<ArrayList<Integer>>();
        for(int i=0;i<numCourses;i++) nodes.add(new ArrayList<>());
        visited = new int[numCourses];
        // establish the graph: add neighbor points
        for(int[] condition:prerequisites){
            int curr = condition[0],pre = condition[1];
            nodes.get(curr).add(pre);
        }
        for(int i = 0;i<numCourses;i++){
            if(visited[i]!=2 && dfs(i)==true)return false;
        }
        return true;
    }
    
    private boolean dfs(int node){
        if(visited[node]==1) return true;
        if(visited[node]==2) return false;
        visited[node] = 1;
        for(int i:nodes.get(node)){
            if(dfs(i)) return false;
        }
        visited[node] = 2;
        return false;
    }
}
